beaker/application.py

```diff
@@ -1,418 +1,963 @@
+                )
+            case LogicSignatureTemplate() as lsig_template:
+                return _lazy_setdefault(
+                    self._precompiled_lsig_templates,
+                    lsig_template,
+                    lambda: PrecompiledLogicSignatureTemplate(lsig_template, client),
                 )
+            case _:
+                raise TypeError("TODO write error message")
+
+    def _register_abi_external(
+        self,
+        method: ABIReturnSubroutine,
+        *,
+        actions: MethodConfigDict,
+        hints: MethodHints,
+        override: bool | None,
+    ) -> None:
+        assert all(cc != CallConfig.NEVER for cc in actions.values())
+        method_sig = method.method_signature()
+        existing_method = self.abi_externals.get(method_sig)
+        if existing_method is None:
+            if override is True:
+                raise ValueError("override=True, but nothing to override")
+        else:
+            if override is False:
+                raise ValueError(
+                    "override=False, but method with matching signature already registered"
+                )
+            # TODO: should we warn if call config differs?
+            self.deregister_abi_method(existing_method.method)
+        self.abi_externals[method_sig] = ABIExternal(
+            actions=actions,
+            method=method,
+            hints=hints,
+        )
 
-                if handler_config.referenced_self:
-                    abi_meth.subroutine.implementation = bound_attr
-
-                self.methods[name] = (abi_meth, handler_config.method_config)
-
-                if handler_config.is_create():
-                    all_creates.append(static_attr)
-                if handler_config.is_update():
-                    all_updates.append(static_attr)
-                if handler_config.is_delete():
-                    all_deletes.append(static_attr)
-                if handler_config.is_opt_in():
-                    all_opt_ins.append(static_attr)
-                if handler_config.is_close_out():
-                    all_close_outs.append(static_attr)
-
-                self.methods[name] = (abi_meth, handler_config.method_config)
-                self.hints[name] = handler_config.hints()
-
-            # Internal subroutines
-            elif handler_config.subroutine is not None:
-                if handler_config.referenced_self:
-                    setattr(self, name, handler_config.subroutine(bound_attr))
-                else:
-                    setattr(
-                        self.__class__,
-                        name,
-                        handler_config.subroutine(static_attr),
+    def deregister_abi_method(
+        self,
+        method_signature_or_reference: str | ABIReturnSubroutine,
+        /,
+    ) -> None:
+        if isinstance(method_signature_or_reference, str):
+            self.abi_externals.pop(method_signature_or_reference)
```

Comment:
legit q, why use pop here and not `del` if we're not using the value returned?

Reply:
Ah, because it was using the returned value and now it's not, good point on just using `del`